### ISSUE TITLE
Duplicate name in a compilation

### ISSUE THREAD

**Issue by paulhuggett** [SNSystems/llvm-project-prepo#40]
Start with something like the following example:

b.c:
~~~C
extern int printf (char const *,...);
void g (void) { printf ("Ipsum\n"); }
~~~

c.c:
~~~C
extern int printf (char const *, ...);
void a (void) { printf ("Ipsum\n"); }
int main () { a (); }
~~~

Makefile:
~~~make
CC = clang  # The "repo clang" should be first on the path
CFLAGS = -O3 -target x86_64-pc-linux-gnu-repo

REPOFILE = clang.db  # Program repository 
OBJECTS = b.o c.o  # Repo ticket files

%.o: %.c
	REPOFILE=$(REPOFILE) clang -c -o $@ $(CFLAGS) $^

.PHONY: all
all: $(OBJECTS)
.PHONY: clean
clean:
	-rm -f $(OBJECTS)
.PHONY: realclean
realclean: clean
	-rm -f $(REPOFILE)
~~~

Build then dump the compilation record for `c.o`:
~~~bash
$ make && pstore-dump --compilation=$(repo-ticket-dump c.o) clang.db
REPOFILE=./clang.db   clang -c -o b.o -O3 -target x86_64-pc-linux-gnu-repo b.c
REPOFILE=./clang.db   clang -c -o c.o -O3 -target x86_64-pc-linux-gnu-repo c.c
---
- file         : 
  compilations : 
      - digest      : 64b87c4d3a8844e85a93c82f370836df
        compilation : 
            members : 
                - digest  : 14a3abc872cf031eab97f9bcfde2d572
                  name    : a
                  linkage : external
                - digest  : 16546f1689a3e586c1c0f53fa730c87f
                  name    : b
                  linkage : external
                - digest  : 561d47bc6c899f9d3214109d63f3a347
                  name    : main
                  linkage : external
                - digest  : 3e3784847c18273a1a9c1e1bffa57e68
                  name    : str
                  linkage : internal
                - digest  : 3e3784847c18273a1a9c1e1bffa57e68
                  name    : str
                  linkage : internal
                - digest  : 4dccddca8e64285a019f6f4db0b3a35d
                  name    : str.2
                  linkage : internal
...
$
~~~

You'll notice that the above (trimmed) `pstore-dump` output contains **two** entries for a member named “str” with internal linkage and a fragment digest of 3e37…7e68. Each name must appear only once in this array.

**Comment by paulhuggett** [comment]
Files are available in a gist (https://gist.github.com/paulhuggett/f4c8ab1d0f81909ce4286db4a39473a1)

**Comment by paulhuggett** [comment]
(I strongly suspect that this either the same problem, or an issue very closely related to, #42.)

**Comment by MaggieYingYi** [comment]
>> 
>> 
>> Could this be related to issue #40?

> No, it is the different issue. I will explain the issue using a small IR code.

> _Originally posted by @MaggieYingYi in https://github.com/SNSystems/llvm-project-prepo/issues/45#issuecomment-528919713_

When I answered the question, I was thinking about issue #42 instead of issue#40, my answer is wrong. Sorry about that. The issue #45 is the same issue as the issue #40, but it is a different from the issue #42. 

However, the current patch for the issue #45 does not fix the issue #40 because the issue #40 involves another issue.

Since the repo doesn’t support the `private` linkage type, the global value `private` linkage type will be converted to the repo `internal` linkage type during the repo object writing stage. This prevents using the existing TicketNode for the dependent GO (with the `private` linkage type).

To solve the issue, use the existing TicketNode for the dependent GO if the dependent GO satisfies the following conditions:
```
1) the dependent GO has the same name and digest as the TicketNode;
2) the TicketNode has an `internal` linkage and the dependent GO has either `private` or `internal` linkage type.
```

**Comment by MaggieYingYi** [comment]
The fix is put into the pull request https://github.com/SNSystems/llvm-project-prepo/pull/47 for the code review.

**Comment by MaggieYingYi** [comment]
Commit c06d3b4 fixed the issue.
